---
title: "ChungProjection"
author: "Melanie"
date: "7/9/2021"
output: html_document
---

```{r libraries}
library(monocle3)
library(CoGAPS)
library(projectR)
library(homologene)
library(biomaRt)
sessionInfo()
```

```{r loadData}
## Load the counts data
counts <- read.table(file = "chungBreast/Chung_Breast_data/GSE75688_GEO_processed_Breast_Cancer_raw_TPM_matrix.txt",
                     header = TRUE)

## Load the metadata
metaData <- read.table(file = "chungBreast/Chung_Breast_data/GSE75688_final_sample_information.txt",
                       header = TRUE)

## Load the gw emt cogaps results
cogapsRes <- readRDS("gwCogapsRes/emtGwCogaps_20.rds")
```

```{r formatData}
## Format the metadata ----
## Remove the pooled samples
metaData <- metaData[grep("Pool", metaData$sample, invert = TRUE, ignore.case = TRUE), ]
## Subset to the tumor samples
metaData <- metaData[grep("Tumor", metaData$index2, ignore.case = TRUE), ]
## Add rownames
rownames(metaData) <- metaData$sample

## Format the counts data ----
## Add a row where duplicated genes get a suffix
counts$unique_gene_name <- make.names(counts$gene_name, unique = TRUE)
## Set unique gene name as the rowname
rownames(counts) <- counts$unique_gene_name
## Extract the gene data
geneData <- counts[, colnames(counts) %in% c("gene_id", "gene_name", "gene_type", "unique_gene_name")]
## colnames
colnames(geneData) <- c("gene_id", "gene_short_name", "gene_type", "unique_gene_name")


## Subset the counts to samples only
counts <- counts[, colnames(counts) %in% rownames(metaData)]

## Check the structure of the counts data
max(counts)
min(counts)

## Check that all the samples are in the data
table(colnames(counts) %in% metaData$sample)
table(metaData$sample %in% colnames(counts) )
dim(counts)
```

```{r projectR}
## Log transform the data and convert to a matrix for project R
transformedMat <- log2(counts + 1)
transformedMat <- as.matrix(transformedMat)

## Duplicate the cogaps res
filteredCogaps <- cogapsRes

filteredCogaps@featureLoadings <- filteredCogaps@featureLoadings[, !colnames(filteredCogaps@featureLoadings) %in%c("Pattern_2", "Pattern_13")]

## Map the gene names across species 
# Use homologene
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl")
map <- getLDS(attributes = c("ensembl_gene_id","hgnc_symbol"),
            filters = "hgnc_symbol", 
            values = geneData$gene_short_name, 
            mart = human,
            attributesL = c("ensembl_gene_id","mgi_symbol","chromosome_name","start_position"),
            martL = mouse)

## merge with the cogaps results
mergedData <- merge(filteredCogaps@featureLoadings, map, by.x = "row.names", 
                    by.y = "MGI.symbol", all.x = TRUE)

## Drop if the HGNC symbol is NA
filteredMergedData <- mergedData[!is.na(mergedData$HGNC.symbol), ]

## Make the human the gene names and remove all non pattern related columns
rownames(filteredMergedData) <- make.names(filteredMergedData$HGNC.symbol, unique = TRUE)

patternMergedData <- filteredMergedData[ , grep("Pattern", colnames(filteredMergedData))]

patternMergedMatrix <- as.matrix(patternMergedData)
## Run projectR
projection <- projectR(data = transformedMat, loadings = patternMergedMatrix)
save(projection, file = "HumanTumorCoGAPSprojection.rdata")
```


```{r makeUmap}
## Add the projection to the dataset
mergedMetaData <- merge(metaData, t(projection),by.x = "sample", by.y = "row.names")
rownames(mergedMetaData) <- mergedMetaData$sample

cds <- new_cell_data_set(expression_data = as.matrix(counts),
                         cell_metadata = mergedMetaData[colnames(counts), ],
                         gene_metadata = geneData)

## Step 1: Normalize and pre-process the data
cds <- preprocess_cds(cds, num_dim = 100)

## Step 3: Reduce the dimensions using UMAP
cds <- reduce_dimension(cds)

pdf("ProjectedPatterns.pdf")
print(plot_cells(cds, color_cells_by = "Pattern_5", cell_size = 1.5))
print(plot_cells(cds, color_cells_by = "Pattern_8", cell_size = 1.5))
print(plot_cells(cds, color_cells_by = "Pattern_14", cell_size = 1.5))
dev.off()


```

